lib/src/revset.rs

```diff
@@ -437,64 +438,162 @@ impl RevsetExpression {
+        // * field values that may match but are mapped to a value that does not should
+        //   be excluded; and
+        //
+        // * field values that may not match but are mapped to a value that does should
+        //   be included.
+        //
+        // We build the expression `(raw_field(pattern) - excludes) | includes`, where:
+        //
+        //     excludes = union_all {
+        //         entry : entry ∈ mailmap,
+        //         (old field absent | old field matches) &
+        //         (new field present & new field doesn’t match)
+        //     }
+        //
+        //     includes = union_all {
+        //         entry : entry ∈ mailmap,
+        //         (old field absent | old field doesn’t match) &
+        //         (new field present & new field matches)
+        //     }
+        //
+        // When matching raw signatures, the `Mailmap` is empty and this reduces to
+        // `raw_field(pattern)`.
+
+        // The first term is a placeholder for later replacement.
+        let mut includes = vec![RevsetExpression::none()];
+        let mut excludes = vec![];
+
+        let pattern_case_insensitive = pattern.to_case_insensitive();
+
+        for entry in mailmap.iter() {
+            let make_old_signature_expr = || {
+                // `.mailmap` entries are always matched case‐insensitively, per
+                // `gitmailmap(5)`.
+                let email_expr: Rc<RevsetExpression> = RevsetExpression::filter(raw_predicate(
+                    SignatureField::Email,
+                    StringPattern::exact_i(entry.old_email()),
+                ));
+                if let Some(name) = entry.old_name() {
+                    email_expr.filtered(raw_predicate(
+                        SignatureField::Name,
+                        StringPattern::exact_i(name),
+                    ))
+                } else {
+                    email_expr
+                }
+            };
+
+            let maybe_new_field_value = match field {
+                SignatureField::Name => entry.new_name(),
+                SignatureField::Email => entry.new_email(),
+            };
+            let Some(new_field_value) = maybe_new_field_value else {
+                continue;
+            };
+            let old_field_value = match field {
+                SignatureField::Name => entry.old_name(),
+                SignatureField::Email => Some(entry.old_email()),
+            };
+
+            // We use `pattern_case_insensitive` for the old field, to respect the
+            // case‐insensitivity of `.mailmap` entries, but `pattern` for the new field, to
+            // determine whether the replacement would actually match.
+            let old_field_matches =
+                old_field_value.map(|field_value| pattern_case_insensitive.matches(field_value));
+            let new_field_matches = pattern.matches(new_field_value);
+
+            if !new_field_matches && old_field_matches.unwrap_or(true) {
```

Comment:
Doesn't `.unwrap_or(true)` mean we're building quite large negative sets? Mailmap entry  doesn't usually contain old name, I suppose. Even if it's more correct, I don't think the added cost is acceptable.

Reply:
FWIW, my original idea was to build query without `excludes`, meaning `author(x)` matches `x` in raw entry or mapped entry. This isn't perfect, but I considered it would be good enough. Mailmap is a bit of a hack.

If we need to guarantee that all shadowed matches are excluded, it's probably better to pass `Mailmap` to `evaluate_revset()` as a resource.